open_spiel/games/dou_dizhu.cc

```diff
@@ -0,0 +1,486 @@
+      return {};
+  }
+}
+
+std::vector<Action> DouDizhuState::DealLegalActions() const {
+  std::vector<Action> legal_actions;
+  legal_actions.reserve(kNumCards - history_.size() + 1);
+
+  if(card_face_up_position_ == -1){
+    for(int i = 0; i < kDealingActionBase; ++i) legal_actions.push_back(i);
+  }else{
+    for (int i = 0; i < kNumCards; ++i) {
+      if (dealer_deck_[i]) legal_actions.push_back(i+kDealingActionBase);
+    }
+  }
+
+
+  return legal_actions;
+}
+
+
+
+std::vector<Action> DouDizhuState::BiddingLegalActions() const {
+  std::vector<Action> legal_actions;
+  legal_actions.reserve(kNumBids + 1);
+  legal_actions.push_back(kPass);
+
+  for (int bid = winning_bid_ + 1; bid <= kNumBids; ++bid) {
+    legal_actions.push_back(kBiddingActionBase + bid);
+  }
+  return legal_actions;
+}
+
+
+
+
+
+
+std::vector<Action> DouDizhuState::PlayLegalActions() const {
+  std::vector<Action> legal_actions;
+  // the leader of a trick must play./ an action and cannot pass
+  if(!new_trick_begin_) legal_actions.push_back(kPass);
+
+
+  std::array<int, kNumRanks> hand = holds_[current_player_];
+  int prev_action = CurrentTrick().WinningAction();
+  SearchForLegalActions(legal_actions, hand, prev_action);
+
+
+  absl::c_sort(legal_actions);
+  return legal_actions;
+}
+
+std::vector<std::pair<Action, double>> DouDizhuState::ChanceOutcomes() const {
+  std::vector<std::pair<Action, double>> outcomes;
+  // int num_cards_remaining = kNumCards - history_.size() + 1;
+
+  int num_cards_remaining = 0;
+  for(int i = 0; i < kNumCards; ++i) num_cards_remaining += dealer_deck_[i];
+  outcomes.reserve(num_cards_remaining);
+
+  if(card_face_up_position_ == -1){
+    for(int i = 0; i < kDealingActionBase; ++i)
+      outcomes.emplace_back(i, 1.0/static_cast<double>(kDealingActionBase));
+  }else{
+    for (int card = 0; card < kNumCards; ++card)
+      if (dealer_deck_[card]) 
+        outcomes.emplace_back(card+kDealingActionBase, 1.0/static_cast<double>(num_cards_remaining));
+  }
+
+
+  return outcomes;
+}
+
+void DouDizhuState::DoApplyAction(Action action) {
+  switch (phase_) {
+    case Phase::kDeal:
+      return ApplyDealAction(action);
+    case Phase::kAuction:
+      return ApplyBiddingAction(action);
+    case Phase::kPlay:
+      return ApplyPlayAction(action);
+    case Phase::kGameOver:
+      SpielFatalError("Cannot act in terminal states");
+  }
+}
+
+void DouDizhuState::ApplyDealAction(int action) {
+  // First decide the face up card
+  if(card_face_up_position_ == -1){
+    card_face_up_position_ = action;
+    return;
+  }
+
+  int dealing_round = static_cast<int>(history_.size()) - 1;
+  // if the current player is dealt the face up card, make it the first one to bid
+  if(dealing_round == history_[0].action){
+    first_player_ = dealing_round % kNumPlayers;
+    card_rank_face_up_ = CardToRank(action-kDealingActionBase);
+  }
+  
+  holds_[((history_.size() - 1) % kNumPlayers)][CardToRank(action-kDealingActionBase)]++;
```

Comment:
Can you use intermediate variables here for the indexing? This line is quite complex.